taf/testlib/virtual_env.py

```diff
@@ -310,17 +376,19 @@ def ensure_public_access(self, try_reuse=False, networks_client=None, routers_cl
 
         _net_cfg.public_network_id = public_network['id']
 
-        router_kwargs = {
-            'routers_client': routers_client,
-            'network_id': _net_cfg.public_network_id,
-            'tenant_id': tenant_id,
-            'enable_snat': True
-        }
-        public_router = self.create_router(**router_kwargs)
-        assert public_router
-        _net_cfg.public_router_id = public_router['id']
+        if create_external_router:
+            router_kwargs = {
+                'routers_client': routers_client,
+                'network_id': _net_cfg.public_network_id,
+                'tenant_id': tenant_id,
+                'enable_snat': True
+            }
+            public_router = self.create_router(**router_kwargs)
+            assert public_router
```

Comment:
Won't the next statement raise some exception if `public_router` is not a dictionary or has no values?

The only thing this statement seems to protect against , beyond the statement that follows it, is an _empty_ `defaultdict`, while it would pass through many other invalid values: `True`, `1`, `[0]`, `object()` and many more.